### ISSUE TITLE
Code completion when editing a macro feels slow

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#54693]
Now that I have had a chance to play with macros, I have noticed that when editing a macro itself, it feels pretty sluggish.

It feels to me like the compilation of the macro is blocking regular code completion and other things. This means any time you edit the macro all errors and code completion features take a few seconds to work.

Would it be possible to offload the compilation to a different isolate? Or is something else going on?

**Comment by scheglov** [comment]
"A few seconds" seems too much. I wonder if it actually crashes and restarts. Do you have anything in `~/.dartServer/.analysis-driver/ex`? I have not tried myself yet.

Yes, running macros is a part of linking summaries, and the analyzer is single threaded.
The current design of the analyzer does not allow us to have multiple analysis operations running.

**Comment by jakemac53** [comment]
I do have some files under there but I am not seeing new ones appear as I continue to make edits. I am also not seeing vscode complain about analyzer crashes.

**Comment by scheglov** [comment]
What specifically do you edit, so that you observe the slowness?

**Comment by jakemac53** [comment]
> What specifically do you edit, so that you observe the slowness?

Really any edit within a macro implementation. So for example when you start writing a macro, I used the quick fix to add the override for the interface I was implementing. Then I go into the body and type `builder.`. It takes a while to get any autocompletions for this.

Similarly, once I finished that line if I add another `builder.` line after that, it takes a while again.

**Comment by scheglov** [comment]
FWIW, I was not able to reproduce it so far, with the previous version of `JsonSerializable`, copied into my `test` package.

**Comment by jakemac53** [comment]
> FWIW, I was not able to reproduce it so far, with the previous version of `JsonSerializable`, copied into my `test` package.

Hmm ok, I was just editing the `observable` macro in the language repo (under working/macros/example).

**Comment by jakemac53** [comment]
Note that it is important to make edits you haven't made before, I noticed that going back to previous states of the file seems to be much faster (I think something is caching results)

**Comment by scheglov** [comment]
True, we do cache both compiled macro implementation in form of kernels, and resulting linked element models.
And I *was* trying to with new code.

I don't see how it could change multi-second delays. If compilation of a macro takes... I don't know, about `100 ms`, and running it then maybe `30 ms`, this is all delay that you should observe.

FWIW, I ran a performance measurement yesterday, mostly to see how much "beautifying" code will cost.
https://dart-review.googlesource.com/c/sdk/+/347500/2/pkg/analyzer/test/src/dart/resolution/macro_test.dart#570
The timings eventually slide under `700 ms`, for running on `100` files, each with unique `UserX` class.
Note that we `useEmptyByteStore()`, so no caching between runs.
We run the series of `100` files multiple times, each time compiling to kernel, and then using `100` times.
And these `700 ms` include both kernel compilation and running over `100` files.
OK, I measured it, compilation takes `120 ms`.
So, you have to pay `120 + 10`, under `150 ms`, not multiple seconds.

Eventually, when everything it hot.

Hot
```
[i: 20]
[KernelCompilationService][timer: 113 ms]
                [time: 642 ms]
```

Cold
```
[KernelCompilationService][timer: 178 ms]
[i: 0]
                [time: 1216 ms]
```

So, my suspicion that it is crash and restart.

**Comment by jakemac53** [comment]
Ok, I do have other macros open which do fail to run, I wonder if it is related to that. I think if in general the expectation is in the low 100s of ms, that should not be noticeable (or at least would be fine, from a UX perspective, even if it is somewhat noticeable).

**Comment by jakemac53** [comment]
Fwiw, with the latest fixes to avoid the crashes I am getting better behavior for sure, but it is still definitely >1s, and I can hear my fan on my computer working hard 🤣 . My guess is we have N separate compilations happening here, one for each keystroke?

[Screen recording 2024-01-29 2.43.22 PM.webm](https://github.com/dart-lang/sdk/assets/984921/3475fa6a-c1c7-4606-b717-b40460d6939b)

**Comment by scheglov** [comment]
No, there should be exactly one compilation, we have a lock on `KernelCompilationService`.
And in general we have single execution flow in the analyzer.

**Comment by jakemac53** [comment]
> No, there should be exactly one compilation, we have a lock on `KernelCompilationService`.

Can we queue up a compilation for each edit possibly though? So it is actually doing back to back compiles for each keystroke? 

> Maybe too old computer? :-)

It is not crazy-fast per core, idk the exact specs (remote machine). It has 128 cores it can just only use one! lol

**Comment by scheglov** [comment]
> > No, there should be exactly one compilation, we have a lock on `KernelCompilationService`.
> 
> Can we queue up a compilation for each edit possibly though? So it is actually doing back to back compiles for each keystroke?

We accumulate file change requests, which will wait until the the current analysis operation is done. Then we apply all changes to files, and take the next analysis operation. Which might cause a kernel compilation. But it is just one compilation for all changes.

> > Maybe too old computer? :-)
> 
> It is not crazy-fast per core, idk the exact specs (remote machine). It has 128 cores it can just only use one! lol

Ah, it might have slow cores though.
That was my experience recently - my Mac is faster per core.

**Comment by jakemac53** [comment]
It is slower per core than my mac for sure - but not that much slower 🤣 .

**Comment by eEQK** [comment]
I had an unused import (of an entire 3rd party library) in a file with a macro that I was editing.

Now that I have removed this import LSP refresh is a lot faster, went down from 8-10s to 0.5s (I'm on AMD 7950x3d)

It looks like importing a ton of code will significantly slow down the LSP.

**Comment by scheglov** [comment]
Might be confusing for users: I changed the macro, but nothing updated.

If taken to extreme, we should not do analysis as you type too.
Want to see errors and warnings? Run `dart analyze` on the command line :-P

**Comment by davidmorgan** [comment]
Yes, there's certainly potential for confusion.

UX is hard ;)

Already working on small examples I've been forced to create a smaller analysis context that excludes users of the macro, because the recompiles are slowing things down too much; I guess we'll want something better before launch.

Agreed with "P2" though, let's solve all the other problems first :)

**Comment by bwilkerson** [comment]
That is theoretically possible, but as Konstantin mentioned elsewhere would require some work to change the way we're currently analyzing code. I don't know how big that effort would be, so I don't know the extent to which that work would impact other efforts. We'd need to sketch out a design first so that we can make some estimates.

**Comment by jakemac53** [comment]
Yep that is fine, and as mentioned earlier I would treat this as a P2, no need to immediately fix as long as we think there is a path to making it better.
